### ISSUE TITLE
Intermittent error removing azuread_application_password

### ISSUE THREAD

**Issue by hbuckle** [hashicorp/terraform-provider-azuread#505]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritise this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritise the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureAD Provider) Version

terraform version: 1.0.3
provider version: 1.6.0

### Affected Resource(s)

* `azuread_application_password`

### Config

```hcl
resource "azuread_application" "azp" {
  display_name     = var.pool_name
  sign_in_audience = "AzureADMyOrg"
  owners = [
    data.azurerm_client_config.current.object_id,
  ]

  web {
    redirect_uris = []
    implicit_grant {
      access_token_issuance_enabled = true
    }
  }
}

resource "azuread_service_principal" "azp" {
  application_id               = azuread_application.azp.application_id
  app_role_assignment_required = false
}

resource "azuread_application_password" "azp" {
  application_object_id = azuread_application.azp.object_id
}
```

### Actual Behavior

I get intermittent errors deleting the azuread_application_password resource

```
Error: Removing password credential "4623c5dc-4b7c-85a3-641b-61d399bd74b0" from application with object ID "16e6c8a1-86c8-43ca-ae59-2cd9dfc8831f"

graphrbac.ApplicationsClient#UpdatePasswordCredentials: Failure responding to
request: StatusCode=404 -- Original Error: autorest/azure: Service returned
an error. Status=404 Code="Unknown" Message="Unknown service error"
Details=[{"odata.error":{"code":"Request_ResourceNotFound","date":"2021-07-28T12:08:09","message":{"lang":"en","value":"Resource
'Application_16e6c8a1-86c8-43ca-ae59-2cd9dfc8831f' does not exist or one of
its queried reference-property objects are not
present."},"requestId":"d258e6f3-4b90-44b2-9366-3d7659bd70d8"}}]
```

The application definitely exists in Azure AD

If I try a few times it eventually works

**Comment by manicminer** [comment]
Hi @hbuckle, thanks for reporting. At a quick glance this looks likely to be an eventual consistency issue. Are you able to try this using the MS Graph beta? If this is a consistency issue I suspect it may be resolved by switching to Microsoft Graph.

When using the legacy API, we have a number of mitigations in place to try and work around these sorts of replication delays, however they don't always work particularly when the API or Azure AD is under heavy load. There are circumstances where the API will report a newly created application as existing despite it not being ready to update - in these cases there isn't much we can do to work around this. Ultimately this problem should go away for the most part when using Microsoft Graph - whilst that API isn't perfect it does make it easier for us to work around this exact problem.

**Comment by dvdvorle** [comment]
I've hit the same issue in the past. And I can confirm using the MS Graph beta seemed to solve it. It also seemed as if this was caused by the simultaneous removal of the `azuread_application_password` and the `azuread_service_principal` resources; Maybe they're both updating the azure_application?

In any case, as a workaround I can offer

``` terraform
resource "azuread_application_password" "azp" {
  application_object_id = azuread_application.azp.object_id
    
  # Just to make sure this won't update the application at the
  # same time during destroy
  depends_on = [azuread_service_principal.azp]
}
```

**Comment by manicminer** [comment]
@dvdvorle Thanks for confirming. In that case I'm going to mark this as resolved since the fix should extend to the recently released version 2.0 which uses MS Graph exclusively.

Note that when you observe resources being provisioned/destroyed simultaneously, that's usually the result of Terraform parallelizing non-dependent operations which is normal.
